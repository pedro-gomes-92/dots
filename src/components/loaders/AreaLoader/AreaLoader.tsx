import React, { FunctionComponent, useState } from 'react';
import classnames from 'classnames';
import { BaseLoader, BaseLoaderProps } from '../BaseLoader';
import { Text } from '../../text';
import { CenterLayout, StackLayout, StackItem } from '../../layouts';
import { IconLoader } from '../IconLoader';
import { Overlay, OverlayProps } from '../../miscellaneous';

export interface AreaLoaderProps extends BaseLoaderProps {
  text?: string;
  hasOverlay?: boolean;
  onOutsideClick?: OverlayProps['onClick'];
}

export const AreaLoader: FunctionComponent<AreaLoaderProps> = ({
  className,
  loader,
  isVisible,
  hasOverlay,
  onOutsideClick,
  ...rest
}: AreaLoaderProps): JSX.Element => {
  const [visible, setVisible] = useState(isVisible);

  let handleOutsideClick: (event: React.MouseEvent<HTMLElement, MouseEvent>) => void;
  if (hasOverlay && onOutsideClick) {
    handleOutsideClick = (event): void => {
      setVisible(false);
      onOutsideClick(event);
    };
  }

  const loaderComponent = (
    <CenterLayout>
      {hasOverlay && <Overlay onClick={handleOutsideClick} />}
      {loader}
    </CenterLayout>
  );

  return (
    <BaseLoader
      className={classnames('is-area', { 'has-overlay': hasOverlay }, className)}
      isVisible={visible}
      loader={loaderComponent}
      {...rest}
    ></BaseLoader>
  );
};

AreaLoader.defaultProps = {
  hasOverlay: false,
  loader: (
    <StackLayout>
      <StackItem alignText="center">
        <IconLoader />
      </StackItem>
      <StackItem>
        <Text text="Loading" />
      </StackItem>
    </StackLayout>
  ),
};
