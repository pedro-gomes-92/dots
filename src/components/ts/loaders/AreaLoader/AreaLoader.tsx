import React, { FunctionComponent, useState } from 'react';
import classnames from 'classnames';
import { BaseLoader, BaseLoaderTypeProps } from '../BaseLoader';
import { Text } from '../../text';
import { CenterLayout, StackLayout, StackItem } from '../../layouts';
import { IconLoader } from '../IconLoader';
import { Overlay, OverlayProps } from '../../miscellaneous';

export interface AreaLoaderProps extends BaseLoaderTypeProps {
  text?: string;
  hasOverlay?: boolean;
  onOutsideClick?: OverlayProps['onClick'];
}

export const AreaLoader: FunctionComponent<AreaLoaderProps> = ({
  className,
  sizeIcon,
  icon,
  isVisible,
  hasOverlay,
  onOutsideClick,
  text,
  ...rest
}: AreaLoaderProps): JSX.Element => {
  const [visible, setVisible] = useState(isVisible);

  let handleOutsideClick;
  if (hasOverlay) {
    handleOutsideClick = (event: React.MouseEvent<HTMLElement, MouseEvent>): void => {
      setVisible(false);
      onOutsideClick(event);
    };
  }

  return (
    <BaseLoader
      className={classnames('is-area', { 'has-overlay': hasOverlay }, className)}
      isVisible={visible}
      {...rest}
    >
      <CenterLayout>
        {hasOverlay && <Overlay onClick={handleOutsideClick} />}
        <StackLayout>
          <StackItem alignText="center">
            <IconLoader name={icon} size={sizeIcon} />
          </StackItem>
          <StackItem>
            <Text text={text} />
          </StackItem>
        </StackLayout>
      </CenterLayout>
    </BaseLoader>
  );
};

AreaLoader.defaultProps = {
  text: 'Loading',
  hasOverlay: false,
  onOutsideClick: (): void => {},
};
