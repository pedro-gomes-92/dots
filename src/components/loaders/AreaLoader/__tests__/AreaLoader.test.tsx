import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { AreaLoader } from '../AreaLoader';

describe('Given <AreaLoader />', (): void => {
  let container: HTMLElement;
  describe('When initializes', (): void => {
    it('Then should match the snapshot', (): void => {
      container = render(<AreaLoader />).container;
      expect(container).toMatchSnapshot();
    });

    describe('And uses hasOverlay', (): void => {
      let overlay: HTMLElement;
      beforeEach((): void => {
        container = render(<AreaLoader hasOverlay />).container;
        overlay = container.querySelector('.overlay');
      });

      it('Then should match the snapshot', (): void => {
        expect(container).toMatchSnapshot();
      });

      it('And should render an overlay', (): void => {
        expect(overlay).toBeTruthy();
      });

      describe('And uses onOutsideClick', (): void => {
        let handleOutsideClick: jest.Mock;

        beforeEach((): void => {
          handleOutsideClick = jest.fn();
          container = render(<AreaLoader hasOverlay onOutsideClick={handleOutsideClick} />).container;
          overlay = container.querySelector('.overlay');

          fireEvent.click(overlay);
        });

        it('Then should trigger onOutsideClick', (): void => {
          expect(handleOutsideClick).toHaveBeenCalledTimes(1);
        });

        it('And should hide the loader', (): void => {
          expect(container.querySelector('.loaders.is-area').classList).toContainEqual('is-hidden');
        });
      });
    });
  });
});
