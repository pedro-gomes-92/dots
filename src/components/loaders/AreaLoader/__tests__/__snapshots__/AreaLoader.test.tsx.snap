// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given <AreaLoader /> When initializes And uses hasOverlay Then should match the snapshot 1`] = `
<div>
  <div
    class="loaders is-area has-overlay"
  >
    <div
      class="layout is-0 layout-center"
    >
      <div
        class="layout-item"
      >
        <div
          class="overlay"
        />
        <div
          class="layout is-0 layout-stack"
        >
          <div
            class="layout-item has-text-alignment-center stack"
          >
            <span
              class="icon is-loader is-relative"
            >
              <i
                class="icon-font icon-font-loader"
              />
            </span>
          </div>
          <div
            class="layout-item stack"
          >
            <span
              class="text"
            >
              Loading
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Given <AreaLoader /> When initializes Then should match the snapshot 1`] = `
<div>
  <div
    class="loaders is-area"
  >
    <div
      class="layout is-0 layout-center"
    >
      <div
        class="layout-item"
      >
        <div
          class="layout is-0 layout-stack"
        >
          <div
            class="layout-item has-text-alignment-center stack"
          >
            <span
              class="icon is-loader is-relative"
            >
              <i
                class="icon-font icon-font-loader"
              />
            </span>
          </div>
          <div
            class="layout-item stack"
          >
            <span
              class="text"
            >
              Loading
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
